prov/verbs/src/verbs_init.c

```diff
@@ -544,6 +548,10 @@ int vrb_find_max_inline(struct ibv_pd *pd, struct ibv_context *context,
 		ibv_destroy_cq(cq);
 	}
 
+	if (pd) {
+		ibv_dealloc_pd(pd);
+	}
```

Comment:
I see that you're following the existing code, but the existing code is goofy.  Assert that create succeed, use the values, then check if succeeded failed prior to calling destroy?

Can you replace the assert(pd) / assert(cq) above with checks, and just return on a failure?  I'm surprised that coverity didn't already complain about the code structure.

Reply:
Thanks for the review! I have replaced the assertion failures with checks as you suggested.

Reply:
So we are not setting init_attr.cap.max_inline_data?  If HW is worth anything it would enforce that when an inject call comes in using inline data part of the WR that's not set when QP was created.
Try this experiment, set # SGE to 1 and do a large inject call (384 bytes) and see what happens.  If it passes, I will be quiet.

Reply:
The function is setting max_inline_data.  It does it repeatedly to find the actual maximum that's supported, which is captured as the ma inject size.  I'm not sure what change you're referring to.

Reply:
Or even worse, this can lead to SQ over-run as each inline WR takes up more room on the queue than indicated in init_attr.

Reply:
The code change does not match commit message.  I still don't see any issues with vrb_get_qp_cap.  The whole thing smells.

Reply:
there shouldn't be a difference between the QP create in vrb_get_qp_cap at getinfo time vs QP created by application at runtime.  If application is indeed using inject_size for max_inline_data (which should be the case).  So why would this change to get by getinfo error but not fail when application creates its QP, again using the same inject size/max_inline_data?

Reply:
@chien-intel Please find below my replies to your concerns/questions:

> And even worse, this can lead to SQ over-run as each inline WR takes up more room on the queue than indicated in init_attr.

I don't quite understand your concern here. The QP is created using the hints provided by the caller or based on the environment variables (FI_VERBS_TX_SIZE, FI_VERBS_TX_IOV_LIMIT, FI_VERBS_INLINE_SIZE, etc...). See `vrb_set_default_info()` for more info.
There is absolutely no intention of overrunning the QP resources set at creation time.

>   I still don't see any issues with vrb_get_qp_cap. The whole thing smells.

I think the real question here is: what's the purpose of `vrb_get_qp_cap()` and why should we keep it? To my understanding, it's a function that allows to validate that the QP can be created with default attributes and max_inline_data. IMHO, testing default attributes is meaningless as the application will likely request different attributes than the default. Finally, the value for max_inline_data is known to be working as `vrb_find_max_inline()` repeatedly creates QPs to discover what's the actual max value supported.

> there shouldn't be a difference between the QP create in vrb_get_qp_cap at getinfo time vs QP created by application at runtime.

There is a difference: an application may want to create a QP at runtime with a lower inline size than the actual maximum in order to create larger TX queues. This is the whole purpose of this PR.

>  If application is indeed using inject_size for max_inline_data (which should be the case)

I don't get your point. Do you mean that applications are expected to use `inject_size = max_inline_data`? If so, this assumption is incorrect. 

> Try this experiment, set # SGE to 1 and do a large inject call (384 bytes) and see what happens. If it passes, I will be quiet.

* Server
```
FI_VERBS_TX_IOV_LIMIT=1 FI_VERBS_RX_IOV_LIMIT=1 fi_rdm_pingpong -p verbs  -j 384 -S 384
bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
384     10k     7.3m        0.06s    118.60       3.24       0.31
```

* Client
```
FI_VERBS_TX_IOV_LIMIT=1 FI_VERBS_RX_IOV_LIMIT=1 fi_rdm_pingpong -p verbs 10.1.1.3 -j 384 -S 384   
bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
384     10k     7.3m        0.06s    118.60       3.24       0.31
```

It works. But to be honest, I don't see why this wouldn't work.

Reply:
> It works. But to be honest, I don't see why this wouldn't work.
Yeah, it will always work, posting 1 WR at a time.  :-)  Posting a queue-depth WRs is a different story.

Here is a realistic hypothetical example to demonstrate SQ overrun.
One key piece of information that I think you are not getting is inline data and SGEs overlap in a WR.  This is why when calculating amount of memory for a SQ, both sizes are used in the calculation.

Let's assume a device that's capable of max 108 bytes of inline data and max 13 SGEs.
Assume 20 bytes header for a SQ operation, this would come out to max 128 bytes for a WR (20 header + 108 inline or 20 header + 13 SGE * 8).

Now take my test case, using your test parameter of 4096 queue depth, 0 inline data and 1 SGE.  WR size may round up to 32 or 64 bytes., we can assume 64 bytes.  Memory allocated to SQ comes out to 64 * 4096 = 262,144 bytes or 256k.  Now, posting queue-depth WR with max inline data size would come out to 4096 * 128 byte WR = 524,288 bytes.  This is the SQ over-run.

fi_getinfo should return valid numbers, period.  Something that the device is capable of supporting which vrb_get_qp_cap is trying to validate.  If application is not interested in sending inline data, perhaps code needs to be modified to use the lowered inline data size and subsequently set inject_size according.

By passing vrb_get_qp_cap instead of root-causing and fixing the code is wrong and subsequently setting inject_size to max inline data size is allowing application to do the wrong thing and making SQ over-run possible.